### ISSUE TITLE
CUDA 12.1

### ISSUE THREAD

**Issue by Andrei-Aksionov** [Lightning-AI/lit-gpt#455]
Hi there 👋 

I've noticed that the latest PyTorch-nightly now can be installed with CUDA 12.1:

<img width="818" alt="Screenshot 2023-08-23 at 2 12 23 PM" src="https://github.com/Lightning-AI/lit-gpt/assets/58434077/be8fd808-a9c1-42b8-8f04-a356daca2c42">

which means that soon enough it will be the version that everyone is using.
Having said that, I think it's worth testing it to see that everything is topnotch:

- [x] Check that training/generation works as before
- [x] Check time/memory for training/generation
- [x] Make sure that bnb is compatible with version 12.1

**Comment by Palmik** [comment]
I tried with Conda and it worked for me. E2E commands:

```
mamba install pytorch torchvision pytorch-cuda=12.1 -c pytorch-nightly -c nvidia
export CUDA_HOME=$CONDA_PREFIX
pip install packaging ninja 
MAX_JOBS=16 pip install 'flash-attn>=2.0.0.post1' --no-build-isolation --no-cache-dir
pip install -r requirements.txt 'bitsandbytes>=0.40.0' scipy 'pillow>=9.4' tokenizers sentencepiece
```

**Comment by Andrei-Aksionov** [comment]
@Palmik Thanks for this!
Could you tell more about what you ran: training, generation, with/without quantization? Have you checked time/memory before and after?
I'll do it myself later on, so it will be good to have results from at least two different people.

**Comment by Andrei-Aksionov** [comment]
There is a significant reason why not to use PyTorch with CUDA v12.1 for now: as per [README](https://github.com/TimDettmers/bitsandbytes#requirements--installation) file from the bitsandbytes repo
> Supported CUDA versions: 10.2 - 12.0

and PyTorch can be installed with either v11.8 or v12.1.

So when to try to run with the latest BNB version (0.41.1) and PyTorch-CUDA12.1 you get:

```python
UserWarning: 
Found duplicate ['libcudart.so', 'libcudart.so.11.0', 'libcudart.so.12.0'] files: 
{PosixPath('/usr/local/cuda/lib64/libcudart.so'), PosixPath('/usr/local/cuda/lib64/libcudart.so.11.0')}.. 
We select the PyTorch default libcudart.so, which is {torch.version.cuda},but this might missmatch 
with the CUDA version that is needed for bitsandbytes.To override this behavior set the 
BNB_CUDA_VERSION=<version string, e.g. 122> environmental variableFor example, if you want to
 use the CUDA version 122BNB_CUDA_VERSION=122 python ...OR set the environmental variable 
in your .bashrc: export BNB_CUDA_VERSION=122In the case of a manual override, make sure you 
set the LD_LIBRARY_PATH, e.g.export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/local/cuda-11.2
```

Of course, we can override the version of libcudart that BNB uses, but I wouldn't want to have such a mismatch between the version that BNB uses and the version that the rest of the code uses. Although the rest of the code will use 12.1 and BNB - 12.0, so it's not that big of a difference.

From the other hand, CUDA v12.x provides some [optimizations and support for the newer architectures](https://developer.nvidia.com/blog/cuda-toolkit-12-0-released-for-general-availability/).

So the question is: should we wait till BNB is updated or not? What do you think @carmocca

**Comment by carmocca** [comment]
The user chooses which cuda version to install (as the README indicates). What exactly are you waiting to change? The default value in the README?

**Comment by Andrei-Aksionov** [comment]
So I did some tests with the code from the main, `PyTorch 2.1` and both versions of CUDA: `11.8` and `12.1`.
For graphics cards I used `T4` and `A10G`. 
Since CUDA, in my experience, is somewhat hardware dependent, the results for other graphic cards, i.e. A100, might be different.

### 1. Nvidia T4 with 16-mixed precision.

| Model       | Loss $_{118}$ | Loss $_{121}$ | Time $_{118}$ | Time $_{121}$ |
|-------------|---------------|---------------|---------------|---------------|
| Pythia-70m  | 2.5641        | 2.5768        | 36.22         | 30.87         |
| Pythia-410m | 1.7997        | 1.7962        | 167.62        | 122.33        |

As we can see on a T4 with the newest version of CUDA the training time is substantially decreased. While debugging I've noticed that 12.1 uses faster `cudaStreamSynchronization` and doesn't use generic kernels at all, so that might explain the speed difference.

| Model       | Token/sec $_{118}$ | Token/sec $_{121}$ |
|-------------|--------------------|--------------------|
| Pythia-70m  | 44.02              | 31.08              |
| Pythia-410m | 31.50              | 26.01              |

But what I didn't expect is the **increase** in generation time: number of tokens per second is much smaller with 12.1 vs 11.8.

### 2. Nvidia T4 with bf16-true precision

| Model       | Loss $_{118}$ | Loss $_{121}$ | Time $_{118}$ | Time $_{121}$ |
|-------------|---------------|---------------|---------------|---------------|
| Pythia-70m  | 3.8046        | 3.8122        | 22.10         | 23.06         |
| Pythia-410m | 1.8418        | 1.8473        | 84.23         | 85.82         |

With A10G, in contrast to T4, the training didn't speed up, in fact now it's slightly slower.

| Model       | Token/sec $_{118}$ | Token/sec $_{121}$ |
|-------------|--------------------|--------------------|
| Pythia-70m  | 119.75             | 62.57              |
| Pythia-410m | 48.21              | 34.67              |

*With 16-mixed precision on A10G the relative difference is the same.*

And again, token generation is much slower with the newest CUDA.

---

For fine-tuning I used LoRA finetuning script with parameters by default and only number of iters was changed to 1k.
Just noticed that for generation `base` script was used, but I think that should not have make a lot of difference.

**Comment by carmocca** [comment]
Those are good numbers @Andrei-Aksionov. I will be updating the generation code at some point with faster variants, so I'm not worried about it at the moment. In my experiments generation also got faster with the newer CUDA version

**Comment by Andrei-Aksionov** [comment]
Awesome.
That means we can close the issue.
